simpleclient/src/main/java/io/prometheus/client/CKMSQuantiles.java

```diff
@@ -35,86 +36,158 @@
  * Implementation of the Cormode, Korn, Muthukrishnan, and Srivastava algorithm
  * for streaming calculation of targeted high-percentile epsilon-approximate
  * quantiles.
- * 
+ * <p>
  * This is a generalization of the earlier work by Greenwald and Khanna (GK),
  * which essentially allows different error bounds on the targeted quantiles,
  * which allows for far more efficient calculation of high-percentiles.
- * 
- * 
+ * <p>
+ * <p>
  * See: Cormode, Korn, Muthukrishnan, and Srivastava
  * "Effective Computation of Biased Quantiles over Data Streams" in ICDE 2005
- * 
+ * <p>
  * Greenwald and Khanna,
  * "Space-efficient online computation of quantile summaries" in SIGMOD 2001
- * 
+ * <p>
+ * <p>
+ * The Approximate Quantiles Algorithm looks like this (Figure 1)
+ * <pre>
+ * Main()
+ *    foreach item v do
+ *      Insert(v);
+ *      if (Compress Condition()) then
+ *          Compress();
+ *
+ * Insert(v):
+ *    r_0:=0;
+ *    for i:=1 to s do
+ *        ri := ri−1 + gi−1;
+ *        if (v < vi) break;
+ *        add (v,1,f(ri,n)−1) to S before vi;
+ *    n++;
+ *
+ * Compress():
+ *    for i := (s−1) downto 1 do
+ *        if (gi + gi+1 + ∆i+1 ≤ f(ri, n)) then
+ *            merge ti and ti+1;
+ *
+ * Output(φ):
+ *    r0:=0;
+ *    for i := 1 to s do
+ *        ri:=ri−1+gi−1;
+ *        if(ri +gi +∆i >φn+f(φn,n)/2)
+ *            print(vi−1); break;
+ * </pre>
  */
-class CKMSQuantiles {
+public final class CKMSQuantiles { /* public for benchmark-module */
     /**
      * Total number of items in stream.
+     * Increases on every insertBatch().
      */
     private int count = 0;
 
     /**
-     * Used for tracking incremental compression.
+     * Current list of sampled items, maintained in sorted order with error
+     * bounds.
+     * <p>
+     * Note: Any algorithm that guarantees to find biased
+     * quantiles φ with error at most φεn in rank must store
+     * Ω(1 min{klog1/φ,log(εn)}) items.
+     * </p>
+     */
+    protected final LinkedList<Item> sample;
+
+    /**
+     * Used for compress condition.
+     * This is a different index than the bufferCount,
+     * because flushing is also done on `quantile.get()`,
+     * but we need to compress, regardless of reaching the bufferCount, to limit space usage.
      */
     private int compressIdx = 0;
 
     /**
-     * Current list of sampled items, maintained in sorted order with error
-     * bounds.
+     * The amount of values observed when to compress, equal to 1/2ε
```

Comment:
I found `equal to 1/2ε` confusing, because it's always 500 independent of ε. I guess it's better to remove that from the comment.

Reply:
Agreed, I experimented with a dynamic epsilon, however, this is tricky as having a 'wide' error would mean that this buffer is filled very quickly. e.g. epsilon of 0.1 gives a size of 1/(0.2) = 5. So then you would add logic stating, at least 500 elems, at most 10000 or something arbitrary like that. 

I will keep it fixed at 500, and add a comment that this is an arbitrary decision.